BDArmory/VesselSpawning/VesselSpawnerBase.cs

```diff
@@ -959,6 +961,166 @@ public void AddToActiveCompetition(Vessel vessel, bool airborne)
                 // Adjust BDACompetitionMode's scoring structures.
                 ContinuousSpawning.Instance.UpdateCompetitionScores(vessel, true);
                 ++ContinuousSpawning.Instance.continuousSpawningScores[vesselName].spawnCount;
+                if (BDACompetitionMode.Instance.competitionIsActive) //these are things that get called on comp start, and such post-comp-start spawning craft will need to have them applied now.
+                {
+                    if (ContinuousSpawning.Instance.vesselsSpawningContinuously && BDACompetitionMode.Instance.competitionIsActive)
```

Comment:
This is already inside checks for both `ContinuousSpawning.Instance.vesselsSpawningContinuously` and `BDACompetitionMode.Instance.competitionIsActive`, however, shouldn't this also apply outside of continuous spawning if a competition is starting or started?

This function could be cleaned up a bit by applying this patch
```diff
diff --git a/BDArmory/VesselSpawning/VesselSpawnerBase.cs b/BDArmory/VesselSpawning/VesselSpawnerBase.cs
index ffc354cbf..f5186aac6 100644
--- a/BDArmory/VesselSpawning/VesselSpawnerBase.cs
+++ b/BDArmory/VesselSpawning/VesselSpawnerBase.cs
@@ -909,2 +909,8 @@ protected IEnumerator PlaceSpawnedVessel(Vessel vessel)
 
+        /// <summary>
+        /// Add a vessel to an active competition.
+        /// Note: this can be called before a competition actually starts, e.g., during the initial spawn of continuous spawn.
+        /// </summary>
+        /// <param name="vessel"></param>
+        /// <param name="airborne"></param>
         public void AddToActiveCompetition(Vessel vessel, bool airborne)
@@ -913,3 +919,4 @@ public void AddToActiveCompetition(Vessel vessel, bool airborne)
             // If a competition is active, update the scoring structure.
-            if ((BDACompetitionMode.Instance.competitionStarting || BDACompetitionMode.Instance.competitionIsActive) && !BDACompetitionMode.Instance.Scores.Players.Contains(vesselName))
+            bool competitionStartingOrStarted = BDACompetitionMode.Instance.competitionStarting || BDACompetitionMode.Instance.competitionIsActive;
+            if (competitionStartingOrStarted && !BDACompetitionMode.Instance.Scores.Players.Contains(vesselName))
             {
@@ -939,16 +946,4 @@ public void AddToActiveCompetition(Vessel vessel, bool airborne)
 
-            //if (!airborne)
-            {
-                weaponManager.AI.ActivatePilot();
-                weaponManager.AI.CommandTakeOff();
-                if (!BDArmorySettings.NO_ENGINES && SpawnUtils.CountActiveEngines(vessel) == 0) // If the vessel didn't activate their engines on AG10, then activate all their engines and hope for the best.
-                {
-                    if (BDArmorySettings.DEBUG_SPAWNING) LogMessage(vessel.vesselName + " didn't activate engines on AG10! Activating ALL their engines.", false);
-                    SpawnUtils.ActivateAllEngines(vessel);
-                }
-                else if (BDArmorySettings.NO_ENGINES && SpawnUtils.CountActiveEngines(vessel) > 0) // Vessel had some active engines. Turn them off if possible.
-                {
-                    SpawnUtils.ActivateAllEngines(vessel, false); //deactivating engines when spawned landed...?
-                }
-            }
+            if (!airborne && competitionStartingOrStarted) AirborneActivation(vessel, false); // Activate ground-spawned craft (air-spawned craft are already active).
+
             // Enable guard mode if a competition is active.
@@ -963,5 +958,4 @@ public void AddToActiveCompetition(Vessel vessel, bool airborne)
                 ++ContinuousSpawning.Instance.continuousSpawningScores[vesselName].spawnCount;
-                if (BDACompetitionMode.Instance.competitionIsActive) //these are things that get called on comp start, and such post-comp-start spawning craft will need to have them applied now.
-                {
-                    if (ContinuousSpawning.Instance.vesselsSpawningContinuously && BDACompetitionMode.Instance.competitionIsActive)
+            }
+            if (competitionStartingOrStarted)
             {
@@ -1122,5 +1116,2 @@ public void AddToActiveCompetition(Vessel vessel, bool airborne)
             }
-                }
-            }

```
and then shifting the details of applying mutators, HOS and RWP tweaks to an appropriate `#region` of `SpawnUtils` so that this section looks like
```csharp
if (competitionStartingOrStarted)
{
    if (BDArmorySettings.MUTATOR_MODE) SpawnUtils.ApplyMutators(vessel, true);
    if (BDArmorySettings.ENABLE_HOS) SpawnUtils.ApplyHOS(vessel, true);
    if (BDArmorySettings.HACK_INTAKES) SpawnUtils.HackIntakes(vessel, true);
    if (BDArmorySettings.RUNWAY_PROJECT) SpawnUtils.ApplyRWP(vessel, true);
}
```
and the same functions can be called in BDACompetitionMode.cs instead of duplicating the code.

Reply:
> however, shouldn't this also apply outside of continuous spawning if a competition is starting or started?

Are there situations where craft would be spawned in after a competition has started that isn't continuous spawn?
...I guess spawning something in manually via vessel mover?
Are there situations where we would want mutators/round specific stuff applied at spawn rather than at comp start? If so, wouldn't it make more sense to just have all HOS/Mutator/RWP stuff applied here, and get rid of the similar calls in the StartDogfightCompetition/SequencedCompetition sections of BDACompetitionMode, as they're now redundant and superfluous?

Reply:
We could always spawn in attack drones on unsuspecting craft as they race through waypoints... and I'm sure there's other situations that we could think of too.